project/src/main/java/com/google/sps/data/PlacesScorer.java

```diff
@@ -0,0 +1,64 @@
+package com.google.sps.data;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import com.google.maps.model.LatLng;
+
+public class PlacesScorer {
+
+    // The coefficients for the scoring algorithm, sum up to 1.
+    private static final double RATING_WEIGHT = 0.7;
+    private static final double DURATION_WEIGHT = 0.3;
+
+    // The maximum possible rating as defined by the Google Places API
+    private static final double MAX_RATING = 5;
+
+    // The maximum durations in seconds,
+    // so that any duration higher than that will not contribute to the place's score.
+    private static final double MAX_DURATION_SECONDS = 40 * 60;
+
+    // A mapping between a place and the driving duration from the place to the user's location.
+    private ImmutableMap<Place, Double> durations;
+
+    // The list of places which their score is calculated.
+    private ImmutableList<Place> places;
+
+    /**
+     * Constructor for PlacesScorer, calculates the durations map used for scoring.
+     *
+     * @param placesToScore A list of places we want to calculate their score
+     * @param userLocation The user's physical location used for duration calculations
+     */
+    public PlacesScorer(ImmutableList<Place> placesToScore, LatLng userLocation) {
+        this.places = placesToScore;
+        this.durations = getDurations(userLocation);
```

Comment:
Re-reading this: you're about to be doing substantial work in the constructor. 

This is problematic as constructor cannot throw exceptions, and your object might be left in an uninitialized state.
Consider adding a factory method which will call some private `init()` function of this class, that will call `getDurations()` post construction.

Please read the following:
https://docs.google.com/document/d/1G3XlK4o52agEF9QNYES22wRLI4PJb4NxM-H9KpR0uVc/edit?usp=sharing&resourcekey=0-9boCwlQ22DjV86NOMGWnLA

Reply:
BTW, have a factory is aligned with the design pattern that we talked about yesterday, in which the `Places` class gets a factory of the `Scorer`.

Reply:
I understand and agree with that point. Anyway I decided not to keep durations as a field, but calculate it only when the scorer is called with getScores - it made everything simpler. So currently there is no need in any special scorer construction.

I did leave the factory class for future use (and a different purpose than discussed here) - in next milestones there will be different implementations of scorers for registered and unregistered users, which will be created by the factory.